packages/app/src/cli/utilities/json-schema.ts

```diff
@@ -0,0 +1,57 @@
+import {FlattenedRemoteSpecification} from '../api/graphql/extension_specifications.js'
+import {BaseConfigType} from '../models/extensions/schemas.js'
+import {RemoteAwareExtensionSpecification} from '../models/extensions/specification.js'
+import {ParseConfigurationResult} from '@shopify/cli-kit/node/schema'
+import {jsonSchemaValidate} from '@shopify/cli-kit/node/json-schema'
+
+/**
+ * Factory returning a function that can parse a configuration object against a locally defined zod schema, and a remotely defined JSON schema based contract
+ * @param merged - The merged specification object from the remote and local sources
+ * @returns A function that can parse a configuration object
+ */
+export function unifiedConfigurationParserFactory(
+  merged: RemoteAwareExtensionSpecification & FlattenedRemoteSpecification,
+) {
+  if (merged.contract === undefined) {
+    return merged.parseConfigurationObject
+  }
+  const contract = JSON.parse(merged.contract)
+
+  const parseConfigurationObject = (config: object): ParseConfigurationResult<BaseConfigType> => {
+    // First we parse with zod. This may also change the format of the data.
+    const zodParse = merged.parseConfigurationObject(config)
```

Comment:
Is this initial parse with zod for cases where you need to get some top-keys from the app toml?
The top-keys are not included in the contract?